omegaconf/grammar_parser.py

```diff
@@ -18,10 +18,10 @@
 
 # Build regex pattern to efficiently identify typical interpolations.
 # See test `test_match_simple_interpolation_pattern` for examples.
-_id = "[a-zA-Z_]\\w*"  # foo, foo_bar, abc123
-_config_key = f"({_id}|\\$)+"  # foo, $bar, $foo$bar$
+_config_key = r"[$\w]+"  # foo, $0, $bar, $foo_$bar123$
 _node_path = f"(\\.)*({_config_key}(\\.{_config_key})*)?"  # foo, .foo.$bar
 _node_inter = f"\\${{\\s*{_node_path}\\s*}}"  # node interpolation ${foo.bar}
+_id = "[a-zA-Z_]\\w*"  # foo, foo_bar, abc123
 _resolver_name = f"({_id}(\\.{_id})*)?"  # foo, ns.bar3, ns_1.ns_2.b0z
 _arg = "[a-zA-Z_0-9/\\-\\+.$%*@]+"  # string representing a resolver argument
```

Comment:
can you explain the change?

Reply:
So the change is with `_config_key` only (`_id` was just moved around).

I haven't dug super deep into it, but I believe the problem was that `_config_key` was ambiguous, because for instance "foo" could be seen as a single `_id`, but also the concatenation of "f" with "oo". So I suspect the pattern matching algorithm was trying lots of different combinations to try and find a match.

I solved it by simply defining `_config_key` from the list of all characters allowed (in this simplified version of the grammar). That means, all standard alpha-numeric characters (`\w` is equivalent to `[a-zA-Z0-9_]` in ASCII mode) plus the $ sign. It has the added benefit of supporting for instance integer IDs (`0`, `12`, ...) which wasn't the case before.